test/test_circularbuffer.jl

```diff
@@ -0,0 +1,34 @@
+using DataStructures
+using Base.Test
+
+
+cb = CircularBuffer{Int}(5)
+@test length(cb) == 0
+@test capacity(cb) == 5
+@test_throws ErrorException first(cb)
+@test isempty(cb) == true
+@test isfull(cb) == false
+
+push!(cb, 1)
+@test length(cb) == 1
+@test capacity(cb) == 5
+@test isfull(cb) == false
+
+append!(cb, 2:8)
+@test length(cb) == capacity(cb)
+@test size(cb) == (length(cb),)
+@test isempty(cb) == false
+@test isfull(cb) == true
+@test convert(Array, cb) == Int[4,5,6,7,8]
+@test cb[1] == 4
+@test cb[2] == 5
+@test cb[3] == 6
+@test cb[4] == 7
+@test cb[5] == 8
+@test_throws ErrorException cb[6]
+@test_throws BoundsError cb[3:6]
```

Comment:
I don't think indexing with ranges/vectors will work on 0.3 unless you implement it yourself, which probably isn't worth it. So maybe encapsulate these in `if VERSION >= v"0.4"`. Or alternatively, I wonder whether people would be opposed to dropping 0.3 support now that 0.4 is out.